tracing/tracing.go

```diff
@@ -0,0 +1,124 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package tracing
+
+import (
+	"context"
+
+	srvconfig "github.com/containerd/containerd/services/server/config"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc"
+)
+
+// InitOpenTelemetry read config and initializes otel middleware, sets the exporter
+// propagator and global tracer provider.
+func InitOpenTelemetry(config *srvconfig.Config) (func(), error) {
+	ctx := context.Background()
+
+	// Check if tracing is configured
+	if config.OpenTelemetry == *new(srvconfig.OpenTelemetryConfig) {
+		logrus.Info("OpenTelemetry configuration is not found in config, tracing will be disabled")
+		return nil, nil
+	}
+
+	// Validate configuration
+	err := config.OpenTelemetry.Validate()
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid open telemetry configuration")
+	}
```

Comment:
(multiple occurrences)
stylistic nit: `err` doesn't need to remain in scope, so you can constrain it to the `if` block

```suggestion
	if err := config.OpenTelemetry.Validate(); err != nil {
		return nil, errors.Wrap(err, "invalid open telemetry configuration")
	}
```

Reply:
Took care of this one.